### ISSUE TITLE
[Bug]: spaces in commands get deleted

### ISSUE THREAD

**Issue by rijobro** [termux/termux-app#2421]
### Problem description

spaces in command get randomly deleted, and I'm not sure why.

### Steps to reproduce the behavior.

`cd ~/Documents` (with space)
> `-bash~/Documents: No such file or directory`

If i then click the upwards arrow, the executed command was
`cd~/Documents` (no space)
So I re-add the space and re-execute:
> `-bash~/Documents: No such file or directory`

This doesn't happen all the time, I'm not sure what triggers it.

### What is the expected behavior?

for spaces not to be deleted

### System information

* Termux application version: 0.117
* Android OS version: 8.0.0
* Device model: samsung s7 edge

**Comment by agnostic-apollo** [comment]
Some program you previously used likely left the terminal in an invalid state. Try long holding terminal -> `More` -> `Reset` when it happens.

Running `reset` or `stty sane` can be helpful for such similar weird behaviour too.

**Comment by rijobro** [comment]
I tried all 3 solutions, none worked. My termux is a fresh installation so doubt it's from a previous program (I only use it to SSH). Can you re-open the issue?

**Comment by agnostic-apollo** [comment]
By previous program I meant, a previous command that you ran in the same terminal. `ssh` can be a likely cause.

Can be a keyboard issue as well. Check https://github.com/termux/termux-app/issues/2385#issuecomment-962669309

Provide reproducible instructions.

**Comment by Grimler91** [comment]
Check if the enforce-char-based-input workaround makes a difference: https://github.com/termux/termux-app/issues/686#issuecomment-681026810

**Comment by rijobro** [comment]
Thanks both for the suggestion. Doesn't seem to solve the issue, but avoiding the `~` character seems to help when on ssh.

**Comment by agnostic-apollo** [comment]
Hopefully, you restarted application or reloading settings after setting value. Post output of `cat ~/.termux/termux.properties | grep enforce-char-based-input`

Also post a video of the bug.

https://f-droid.org/en/packages/com.orpheusdroid.screenrecorder/

**Comment by rijobro** [comment]
~ $ cat .termux/termux.properties
enforce-char-based-input=true

**Comment by agnostic-apollo** [comment]
Thanks. But without exact reproducible command samples, the bug cannot be fixed if its a termux app issue.

**Comment by Grimler91** [comment]
You can also test installing another keyboard, like hackers keyboard. If it works with that one then we know that issue is with samsung's stock (or do you have a custom rom installed?) keyboard

**Comment by bjlockie** [comment]
Press end after you add the space.
Pressing enter in the middle deletes the added space.
Enter at the end of the command works for me.

**Comment by Gavin-Paolucci-Kleinow** [comment]
I'm experiencing the same problem, and hitting end or moving to the end of the line just causes a different problem--it cuts off the last character. I'm able to reproduce the problems. Here's a sample of what happens:

'' ~ $ test=5''
First command where the space is missed
~ $ echo$test
No command echo5 found, did you mean:
 Command echo in package coreutils
Trying just adding a space between the o and $ and hitting enter immediately but the result is the same command as before
~ $ echo$test
No command echo5 found, did you mean:
 Command echo in package coreutils
Trying adding the space and hitting end before enter, now the last character is removed 
~ $ echo $tes
Trying adding the space and moving the cursor to the end of the line with the arrow keys before hitting enter, same command ran with the final character removed. 
~ $ echo $tes

**Comment by bjlockie** [comment]
I press the end key to go to the end of the line.
I wonder if that is different than using the cursor key.

**Comment by Gavin-Paolucci-Kleinow** [comment]
Setting `enforce-char-based-input = true` in termux.properties seems to have solved the problem for me with SwiftKey.

**Comment by SDRausty** [comment]
GoogleChrome Browser refreshed my text;   Hence I am writing this text once more. 

[Bug]: .... get deleted #2421

I advertized https://github.com/BuildAPKs/buildAPKs.keyboards/blob/main/ma.bash as a potential solution for https://github.com/termux/termux-app/issues/686 and elswhere.  Some of these keyboards do more than delete spaces now.  Also, the amount of keyboards which build in native Termux with Termux utilities has fallen significantly:
![Screenshot_20220125-225128](https://user-images.githubusercontent.com/27742457/151101978-9cac5164-eaba-4384-8814-e91184af6b16.png)

My contributions in https://github.com/termux/termux-app/issues/2421 are removed;  Please reinstate all of them.

**Comment by SDRausty** [comment]
Google Chrome Browser refreshed my text;   Hence I am writing this text once more from scratch:

[Bug]: .... get deleted #2421

I advertized https://github.com/BuildAPKs/buildAPKs.keyboards/blob/main/ma.bash as a potential solution for https://github.com/termux/termux-app/issues/686 and elswhere.  Some of these keyboards do more than delete spaces now.  Also, the amount of keyboards which build in native Termux with Termux utilities has fallen significantly:
![Screenshot_20220125-225128](https://user-images.githubusercontent.com/27742457/151101978-9cac5164-eaba-4384-8814-e91184af6b16.png)

My contributions in https://github.com/termux/termux-app/issues/686 are removed;  Please reinstate all of them.

**Comment by SDRausty** [comment]
Google Chrome Browser refreshed my text;   Hence I am writing this text once more from scratch:

[Bug]: .... get deleted #2421

I advertized https://github.com/BuildAPKs/buildAPKs.keyboards/blob/main/ma.bash as a potential solution for https://github.com/termux/termux-app/issues/686 and elswhere.  Some of these keyboards do more than delete spaces now.  Also, the amount of keyboards which build in native Termux with Termux utilities has fallen significantly:
![Screenshot_20220125-225128](https://user-images.githubusercontent.com/27742457/151101978-9cac5164-eaba-4384-8814-e91184af6b16.png)

My contributions in https://github.com/termux/termux-app/issues/686 are removed;  Please reinstate all of them.

**Comment by SDRausty** [comment]
One of the reasons that I am suggesting [buildAPKs.keyboards](https://github.com/BuildAPKs/buildAPKs.keyboards/) as assistance in resolving this issue is that with at present the variety of Android character keyboards which build directly in Termux that [buildAPKs.keyboards](https://github.com/BuildAPKs/buildAPKs.keyboards/blob/main/ma.bash) has to offer, the 'spaces in commands get deleted' issue varies between the different keyboards used; i.e. As various Android touch keyboards have at times, and from what I see sometimes consistently erroneous results, this volume of functioning code can be used to make Termux app more robust at intercepting keyboard codes in order to function as intended:
```
find ~/buildAPKs/sources/keyboards/ -name *.java -or -name *.xml -exec cat {} \; | wc
108393  298009 4342055
```
This collection is currently 298009 ewords long.

**Comment by agnostic-apollo** [comment]
>Same here! My keyboard app is SwiftKey too.

Do you mean setting the property fixed the issue? There might be some additional settings in the app's keyboard settings that could be affecting behaviour.

Also the video is private.

**Comment by kirisakow** [comment]
> Do you mean setting the property fixed the issue? 

It has.

> There might be some additional settings in the app's keyboard settings that could be affecting behaviour.

It might. 

> Also the video is private.

The video is publicly accessible. I double checked just now

**Comment by HugeFrog24** [comment]
Agreed. Another one of those bugs that’s impossible to search for because you don’t know how to describe it. Characters vanishing when editing commands in Termux with Microsoft SwiftKey. Try Googling that. Good luck.

**Suggested fix (thanks to @kirisakow):**
1. Uninstalled Play Store version of Termux (in my case, it didn’t respect `termux.properties`).
2. Installed latest from GitHub: [v0.118.2](https://github.com/termux/termux-app/releases/latest)
3. Edited `~/.termux/termux.properties`, uncommented:
```
enforce-char-based-input = true
```
4. Ran `termux-reload-settings` and restarted Termux.

Done. No more broken input. Spaces stay where they should. Sanity restored.
